maestro-client/src/main/java/maestro/ViewHierarchy.kt

```diff
@@ -89,3 +95,27 @@ value class ViewHierarchy(val root: TreeNode) {
         return root.aggregate()
     }
 }
+
+fun TreeNode.filterOutOfBounds(width: Int, height: Int): TreeNode? {
+    val element = toUiElement()
+    val isXWithinBounds = element.bounds.x in 0..width
+    val isYWithinBounds = element.bounds.y in 0..height
```

Comment:
You are checking for top left corner, however it could be that it is partially out of bounds while the bottom of the view is still in the viewport and could be visible

Reply:
We'd need something along the lines of:
```
let isOverlapping = abs(screen.origin.x - view.origin.x) < 1 &&
                   abs(screen.origin.y - view.origin.y) < 1 &&
                   abs(screen.size.width - view.size.width) < 1 &&
                   abs(screen.size.height - view.size.height) < 1
```

Reply:
>You are checking for top left corner, however it could be that it is partially out of bounds while the bottom of the view is still in the viewport and could be visible

Should we consider partial elements as "visible"? We could check if any of the corners are within bounds and consider it "visible".

Reply:
_Or_ we could say that the view is visible if >10% of its area is in viewport, just to remove edge cases where only a small piece of the view is on the screen.

Reply:
assertVisible and scrollToVisible should have the same **sensible** definition. Something to consider when going for a more or less heuristic definition like '>10%'.

Reply:
Agree, if we apply that criteria it should apply to both

Reply:
<img width="448" alt="Screenshot 2023-01-31 at 18 14 54" src="https://user-images.githubusercontent.com/2115575/215784355-b24647b3-4b88-4939-b7d0-8d3fad929867.png">

Here's a valid case:

Both texts (top/bottom) that you see cut off are considered 100% visible since _technically_ they are within screen bounds. However most users would not consider this visible since the browser is hiding the texts.

A "hack" i did for scrollUntilVisible was to add a 10% element padding. Not sure we want the same for assertVisible.

Any thoughts?

Reply:
I would actually consider those to be visible technically speaking. As in, I would not expect my `assertVisible` to fail

Reply:
Right, for `assertVisible` it makes sense. For `scrollUntilVisible` i'd like to add a bit more control when the scroll stops (in another PR).

Conclusion for this PR. Consider elements visible, if 10% is visible, otherwise they remove from hierarchy.

(Will add tests for this)

Reply:
That makes sense